scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamFunctions.scala

```diff
@@ -58,4 +62,35 @@ object BeamFunctions {
       }
     }
   }
+
+  case class HashJoinFn[K, V, U, W](
+    joiner: (K, V, Iterable[U]) => Iterator[W],
+    sideInput: PCollectionView[java.util.Map[K, java.lang.Iterable[U]]]
+  ) extends DoFn[(K, V), (K, W)] {
+    private[this] var mapRight: java.util.Map[K, java.lang.Iterable[U]] = null
+    private[this] val emptyUs: Iterable[U] = Seq.empty[U]
+
+    @ProcessElement
+    def processElement(c: DoFn[(K, V), (K, W)]#ProcessContext): Unit = {
+      if (mapRight == null) {
+        mapRight = c.sideInput(sideInput)
+      }
+      val key = c.element()._1
+      val value = c.element()._2
+      val it = mapRight.get(key) match {
```

Comment:
nit: `.getOrElse(key, emptyUs)` or something like that

Reply:
I suggested this. This code compiles to an if/else. The getOrElse has to allocate a closure to model the lazy nature of getOrElse. Then the else case is a function call, not the other side of a branch.

Inside the hot path of the backends I would use pretty low level scala.